File: scripts/export_data.py
"""导出结构化数据为 CSV / JSON / Markdown。

用法：
  python scripts/export_data.py --format json
  python scripts/export_data.py --format csv --schema-type NewsItem --since 2026-09-01 --out out.csv

只读打开数据库（mode=ro），不影响在跑的采集。
"""
from __future__ import annotations

import argparse
import csv
import io
import json
import pathlib
import sqlite3
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[1]))

from core.dotenv import load_dotenv

load_dotenv()

REPO_ROOT = pathlib.Path(__file__).resolve().parents[1]
_META_FIELDS = ["id", "run_id", "source_url", "schema_type", "created_at"]


def fetch_rows(db_path: pathlib.Path, schema_type: str | None = None,
               since: str | None = None, until: str | None = None) -> list[dict]:
    """从 extracted_items 读取并解析 content；数据库只读打开。"""
    if not db_path.exists():
        raise FileNotFoundError(f"数据库不存在：{db_path}（先运行 run_once / run_daemon）")
    conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
    conn.row_factory = sqlite3.Row
    try:
        sql = ("SELECT id, run_id, source_url, schema_type, content, created_at "
               "FROM extracted_items WHERE 1 = 1")
        params: list = []
        if schema_type:
            sql += " AND schema_type = ?"
            params.append(schema_type)
        if since:
            sql += " AND date(created_at) >= date(?)"
            params.append(since)
        if until:
            sql += " AND date(created_at) <= date(?)"
            params.append(until)
        sql += " ORDER BY id"
        return [{"id": r["id"], "run_id": r["run_id"], "source_url": r["source_url"],
                 "schema_type": r["schema_type"], "created_at": r["created_at"],
                 "item": json.loads(r["content"])}
                for r in conn.execute(sql, params)]
    finally:
        conn.close()


def to_json(rows: list[dict]) -> str:
    return json.dumps({"count": len(rows), "items": rows},
                      ensure_ascii=False, indent=2)


def to_csv(rows: list[dict]) -> str:
    """元数据列 + item 字段并集；嵌套值序列化为 JSON 字符串。"""
    item_keys: list[str] = []
    for row in rows:
        for key in row["item"]:
            if key not in item_keys:
                item_keys.append(key)
    buf = io.StringIO()
    writer = csv.writer(buf, lineterminator="\n")
    writer.writerow(_META_FIELDS + item_keys)
    for row in rows:
        meta = [row[f] for f in _META_FIELDS]
        cells = []
        for key in item_keys:
            value = row["item"].get(key)
            if isinstance(value, (list, dict)):
                value = json.dumps(value, ensure_ascii=False)
            cells.append(value)
        writer.writerow(meta + cells)
    # utf-8-sig BOM：Excel 打开中文不乱码
    return "\ufeff" + buf.getvalue()


def to_markdown(rows: list[dict]) -> str:
    lines: list[str] = [f"# 采集数据导出（{len(rows)} 条）", ""]
    for row in rows:
        item = row["item"]
        title = item.get("title") or item.get("headline") or f"条目 #{row['id']}"
        lines.append(f"## {title}")
        lines.append(f"- 来源：{row['source_url']}")
        lines.append(f"- 类型：{row['schema_type']}｜采集时间：{row['created_at']}")
        for key, value in item.items():
            if key in ("title", "headline"):
                continue
            if isinstance(value, (list, dict)):
                value = json.dumps(value, ensure_ascii=False)
            lines.append(f"- {key}：{value}")
        lines.append("")
    return "\n".join(lines)


def main() -> int:
    parser = argparse.ArgumentParser(description="导出 extracted_items 为 CSV/JSON/Markdown")
    parser.add_argument("--format", choices=["csv", "json", "markdown"], default="json")
    parser.add_argument("--schema-type", default=None, help="按 Schema 类名过滤")
    parser.add_argument("--since", default=None, help="起始日期 YYYY-MM-DD（含）")
    parser.add_argument("--until", default=None, help="结束日期 YYYY-MM-DD（含）")
    parser.add_argument("--db", default=str(REPO_ROOT / "data" / "collector.db"))
    parser.add_argument("--out", default=None, help="输出文件路径（默认 stdout）")
    args = parser.parse_args()

    try:
        rows = fetch_rows(pathlib.Path(args.db), args.schema_type, args.since, args.until)
    except (FileNotFoundError, sqlite3.OperationalError, json.JSONDecodeError) as e:
        print(f"导出失败：{e}", file=sys.stderr)
        return 2

    output = {"csv": to_csv, "json": to_json, "markdown": to_markdown}[args.format](rows)
    if args.out:
        out = pathlib.Path(args.out)
        if args.format == "csv":
            out.write_text(output, encoding="utf-8")  # 已含 BOM 字符
        else:
            out.write_text(output, encoding="utf-8")
        print(f"已导出 {len(rows)} 条 → {out}", file=sys.stderr)
    else:
        sys.stdout.write(output)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/export-data.ts
/** 导出结构化数据为 CSV / JSON / Markdown。
 *  用法：node scripts/export-data.ts --format json [--schema-type NewsItem] [--since 2026-09-01] [--out out.json] */
import { resolve } from "node:path";
import { writeFileSync } from "node:fs";

import { Database } from "../src/storage/db.ts";

const REPO_ROOT = resolve(import.meta.dirname, "..");
const META_FIELDS = ["id", "run_id", "source_url", "schema_type", "created_at"];

interface ExportRow {
  id: number;
  run_id: number;
  source_url: string;
  schema_type: string;
  created_at: string;
  item: Record<string, unknown>;
}

export function fetchRows(
  db: Database,
  opts: { schemaType?: string; since?: string; until?: string } = {},
): ExportRow[] {
  let sql = "SELECT id, run_id, source_url, schema_type, content, created_at FROM extracted_items WHERE 1 = 1";
  const params: string[] = [];
  if (opts.schemaType) {
    sql += " AND schema_type = ?";
    params.push(opts.schemaType);
  }
  if (opts.since) {
    sql += " AND date(created_at) >= date(?)";
    params.push(opts.since);
  }
  if (opts.until) {
    sql += " AND date(created_at) <= date(?)";
    params.push(opts.until);
  }
  sql += " ORDER BY id";
  return (db.conn.prepare(sql).all(...params) as Record<string, unknown>[]).map((r) => ({
    id: r.id as number,
    run_id: r.run_id as number,
    source_url: r.source_url as string,
    schema_type: r.schema_type as string,
    created_at: r.created_at as string,
    item: JSON.parse(r.content as string),
  }));
}

export function toJson(rows: ExportRow[]): string {
  return JSON.stringify({ count: rows.length, items: rows }, null, 2);
}

export function toCsv(rows: ExportRow[]): string {
  const itemKeys: string[] = [];
  for (const row of rows) {
    for (const key of Object.keys(row.item)) {
      if (!itemKeys.includes(key)) itemKeys.push(key);
    }
  }
  const esc = (v: unknown): string => {
    if (v === null || v === undefined) return "";
    const s = typeof v === "object" ? JSON.stringify(v) : String(v);
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const lines = [META_FIELDS.concat(itemKeys).join(",")];
  for (const row of rows) {
    lines.push(
      META_FIELDS.map((f) => esc(row[f as keyof ExportRow]))
        .concat(itemKeys.map((k) => esc(row.item[k])))
        .join(","),
    );
  }
  // utf-8-sig BOM：Excel 打开中文不乱码
  return "﻿" + lines.join("\n") + "\n";
}

export function toMarkdown(rows: ExportRow[]): string {
  const lines = [`# 采集数据导出（${rows.length} 条）`, ""];
  for (const row of rows) {
    const item = row.item;
    const title = (item.title as string) || (item.headline as string) || `条目 #${row.id}`;
    lines.push(`## ${title}`);
    lines.push(`- 来源：${row.source_url}`);
    lines.push(`- 类型：${row.schema_type}｜采集时间：${row.created_at}`);
    for (const [key, value] of Object.entries(item)) {
      if (key === "title" || key === "headline") continue;
      const text = typeof value === "object" ? JSON.stringify(value) : String(value);
      lines.push(`- ${key}：${text}`);
    }
    lines.push("");
  }
  return lines.join("\n");
}

function parseArgs(argv: string[]): Record<string, string | boolean> {
  const args: Record<string, string | boolean> = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a.startsWith("--")) {
      const key = a.slice(2);
      const next = argv[i + 1];
      if (next && !next.startsWith("--")) {
        args[key] = next;
        i++;
      } else args[key] = true;
    }
  }
  return args;
}

function main(): number {
  const args = parseArgs(process.argv.slice(2));
  const format = (args.format as string) ?? "json";
  const dbPath = (args.db as string) ?? resolve(REPO_ROOT, "data/collector.db");
  let db: Database;
  try {
    db = new Database(dbPath);
  } catch (e) {
    console.error(`导出失败：数据库不存在或无法打开（${dbPath}）`);
    return 2;
  }
  try {
    const rows = fetchRows(db, {
      schemaType: args["schema-type"] as string | undefined,
      since: args.since as string | undefined,
      until: args.until as string | undefined,
    });
    const renderers: Record<string, (r: ExportRow[]) => string> = {
      csv: toCsv,
      json: toJson,
      markdown: toMarkdown,
    };
    const render = renderers[format];
    if (!render) {
      console.error(`未知格式: ${format}（可选 csv / json / markdown）`);
      return 2;
    }
    const output = render(rows);
    if (args.out) {
      writeFileSync(args.out as string, output, "utf8");
      console.error(`已导出 ${rows.length} 条 → ${args.out}`);
    } else {
      process.stdout.write(output);
    }
    return 0;
  } finally {
    db.close();
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exit(main());
}
